azure/services/securitygroups/securitygroups.go

```diff
@@ -83,7 +95,55 @@ func (s *Service) Reconcile(ctx context.Context) error {
 	// We go through the list of security groups to reconcile each one, independently of the result of the previous one.
 	// If multiple errors occur, we return the most pressing one.
 	//  Order of precedence (highest -> lowest) is: error that is not an operationNotDoneError (i.e. error creating) -> operationNotDoneError (i.e. creating in progress) -> no error (i.e. created)
-	for _, nsgSpec := range specs {
+	for _, resourceSpec := range specs {
+		nsgSpec, ok := resourceSpec.(*NSGSpec)
+		if !ok {
+			return errors.New("cannot convert network security group spec")
+		}
+
+		newAnnotation := map[string]interface{}{}
+
+		// We only want to update annotations for the security rules that are always managed by CAPZ.
+		for _, securityRuleSpec := range nsgSpec.SecurityRulesSpecs {
+			if _, alwaysManaged := securityRuleLastAppliedAnnotation[securityRuleSpec.Annotation]; !alwaysManaged {
+				continue
+			}
+			newAnnotation[securityRuleSpec.SecurityRule.Name] = securityRuleSpec.SecurityRule
+		}
+
+		// Retrieve the last applied security rules for all NSGs.
+		lastAppliedSecurityRulesAll, err := s.Scope.AnnotationJSON(azure.SecurityRuleLastAppliedAnnotation)
```

Comment:
I'm keeping this in the loop because the lastAppliedSecurityRulesAll gets updated in between iterations, and is in turn used to update the annotation. In order to keep it up to date, I kept this call within the loop.

Reply:
Ah yep I see that now.